### PR TITLE
Minor fixes for local allocations

### PR DESCRIPTION
**Pull request by stedolan**
Before the larger mode variables patch, here's a couple of independent fixes:

  - Fix a bug in Comballoc, where the two independent allocation sequences were ended simultaneously instead of independently
  - Fix a bug in Typecore, where the intermediate return modes of curried lambdas were ignored
  - Made the GC support slightly less bad by scanning local allocs at GC time (still needs work, this is just enough to fix some crashes)
  - Print total local allocs if OCAMLRUNPARAM=v=0x400 is set. (This may be removed later, as there may be a perf penalty. It's useful for debugging now, though)

### CODE REVIEW

File: runtime/memory.c [hunk 1]
```diff
@@ -706,6 +707,7 @@ void caml_local_realloc()
   while (Caml_state->local_sp < -new_bsize) new_bsize *= 2;
   stkbase = caml_stat_alloc(new_bsize + sizeof(struct region_stack));
   stk = (struct region_stack*)(stkbase + new_bsize);
+  memset(stkbase, 0x42, new_bsize);
```
Comment (lpw25) [inline runtime/memory.c]: Is this setting newly allocated memory to a known value? Presumably we only want to do this in the debug runtime for the release. Could you leave a comment to make sure we don't forget to remove it.

File: typing/typecore.ml [hunk 2]
```diff
@@ -3814,11 +3814,15 @@ and type_function ?in_function loc attrs env mode ty_expected_explained l caseli
                                           ty_fun,
                                           explanation)))
   in
-  if not curry then
-    begin match Types.Alloc_mode.submode ret_mode Alloc_heap with
+  if curry then begin
+    match Types.Alloc_mode.submode mode ret_mode with
```
Comment (lpw25) [inline typing/typecore.ml:3818]: This change is interesting. I guess it captures the idea that partially applying a function captures both the argument and the original function. This means that code like this isn't allowed:
```ocaml
let bar z =
  let[@stack]foo : int -> int -> int = fun x y -> x + y + z in
  ...
```
and it must instead be written as:
```ocaml
let bar z =
  let[@stack]foo : int -> (int -> int)[@stackret] = fun x y -> x + y + z in
  ...
```
I think this is another argument in favour of the proper currying support in types.

**Review by lpw25 — APPROVED** [review]

File: runtime/memory.c (hunk 1 above)
Comment (stedolan) [reply runtime/memory.c]: Added a FIXME. This is all going to change in a week or two anyway when I sort out GC support properly. (This is just to make it not segfault for now)

File: typing/typecore.ml (hunk 2 above)
Comment (stedolan) [reply typing/typecore.ml:3818]: Agreed!

**Pull request merged**